### DISCUSSION TITLE
Replace SQLite Libraries with Berkeley DB Libraries

### DISCUSSION THREAD

**Discussion by phreed** [sqlitebrowser/sqlitebrowser#2995] [Q&A]
The Oracle forum calls out the sqlitebrowser as a tool for editing Berkeley DB.
https://community.oracle.com/tech/developers/discussion/3561584/editor-for-berkeley-db

Specifically, I need to edit my BDB backed Subversion repository before migrating to Git.
Here is what they say:
> You will need to be using the SQL API for BDB for these to work.   
> If you find ones that will work for SQLite then they will work for BDB.
> To use this browser you will need to replace the sqlite libraries with the BDB libraries.
> The steps to do this are contained in the BDB docs.

**Comment by phreed** [comment]
I am looking through the BDB docs some  "steps" are found in...
https://www.oracle.com/a/ocom/docs/bdb-sqlite-comparison-wp-176431.pdf

> Berkeley DB’s SQL API is compatible with SQLite. 
> Thus it enables it to act as a drop-in replacement for SQLite.
> That is, you add a single header file (db.h in addition to your existing sqlite3.h) and instead of linking the SQLite
> library you link against two Berkeley DB libraries: db-5 and db_sql-5. 
> Recompile, re-link, and you are in business.
> Your application that once used SQLite has been ported to use Berkeley DB.

**Comment by chrisjlocke** [comment]
What operating system are you using?
You can't simply 'replace' the SQLite library with the Berkley libraries - DB4S has been written to use SQLite and it expects things to work that way.  The documentation mentions the libraries are quite similar though, so it might be worth a punt.  Think its a bit like looking at a Ford Mondeo and a Tesla 3 and while they might look similar sized cars, you can't just switch engines and expect everything to work...

There is a command-line tool for Berkley DB though, so do you need a GUI tool?  If you're doing mild amendments, the command-line might be enough?

**Comment by justinclift** [comment]
Skimming through that pdf quickly, it says:

> Oracle has taken the SQLite source code, extracted everything from its storage layer up and grafted it on BDB.

That being the case, it really *might* work to drop a compiled version (eg .dll) of this BDB/SQLite thing into the DB4S folder, replacing the SQLite .dll file.

If that doesn't work, then compiling DB4S against this BDB/SQLite thing might do the job.  It *sounds* like there's a reasonable chance of success anyway.

Further along in the PDF (page 4, I'm really only just skimming instead of proper reading though), it has this:

> Berkeley DB’s SQL API is compatible with SQLite. Thus it enables it to act as a drop-in replacement for SQLite.
> That is, you add a single header file (db.h in addition to your existing sqlite3.h) and instead of linking the SQLite
> library you link against two Berkeley DB libraries: db-5 and db_sql-5. Recompile, re-link, and you are in business.
> Your application that once used SQLite has been ported to use Berkeley DB.

So yeah, it does sound like recompiling would be needed.
